tests/techmap/constmap.ys

```diff
@@ -0,0 +1,40 @@
+read_verilog << EOT
+
+module test();
+  wire [31:0] in;
+  wire [31:0] out;
+  assign out = in + 16;
+endmodule
+
+EOT
+
+constmap -cell const_cell O value
+select -assert-count 1 t:const_cell
+select -assert-count 1 r:value=16
+
+design -reset
+
+read_verilog -lib << EOT
+module const_cell(O);
+output O;
+endmodule
+EOT
+
+read_verilog << EOT
+
+module test();
+  wire [31:0] in;
+  wire [31:0] out1;
+  wire [31:0] out2;
+  assign out1 = in + 16;
+  assign out2 = in + 32;
+endmodule
+
+EOT
+
+constmap -cell const_cell O value
```

Comment:
For this case where there is a (blackbox) module provided, I wonder if it would make sense to test whether the module has the parameter being assigned.  Also because if you call `hierarchy -check` after running this you get an error:
```
ERROR: Module `const_cell' referenced in module `test' in cell `$auto$constmap.cc:36:constmap_worker$9' does not have a parameter named 'value'.
```
And most (all?) of the `synth_*` script passes run `hierarchy -check` before writing out the output so it would probably be expected for this pass to not fail there.

Reply:
I think it is good that `hierarchy -check` fails should be expected since this tells the user they have used something that is not part of the primitive lib.

Reply:
That's not what `hierarchy -check` is for though.  `hierarchy -simcheck` has the added constraint of failing on blackbox instances which I think is the situation you are meaning.  If you read the `const_cell` module with `-noblackbox` instead of `-lib`, giving you a non-blackbox module without an implementation, that *should* fail `hierarchy -check`.  As should the first case where the `const_cell` module has no definition.  Adding the parameter (like you've now done) stops the error from appearing, but `const_cell` still isn't part of the primitive lib.

Reply:
So, should I add an assert at the pass to check if the cell is a known module and fail when the parameter is unavailable?

Reply:
Not an assert, it would be a `log_cmd_error` to indicate that the command is invalid for the current design/options/selection.  Or a `log_error` if you've already modified the design.  In a script the two are identical, but in an interactive prompt `log_cmd_error` aborts just the command instead of Yosys.